### ISSUE TITLE
Compilation error with Vex.Struct  validates by

### ISSUE THREAD

**Issue by bitchef** [CargoSense/vex#11]
Hi,
I am a new Vex user and I am loving it (^_^)b Recently, I stumbled on the following error while trying to vaiidate a structure attribute with a custom function:

Environment: Elixir 1.0.3  -  Erlang/OTP 17 
```
** (ArgumentError) cannot inject attribute @vex_validations into function/macro because cannot escape #Function<1.98856283/1 in :elixir_compiler_5.__MODULE__/1>. The supported values are: lists, tuples, maps, atoms, numbers, bitstrings, pids and remote functions in the format &Mod.fun/arity
    (elixir) lib/kernel.ex:2043: Kernel.do_at/4
    (elixir) expanding macro: Kernel.@/1
    test/test_helper.exs:8: UserTest.__vex_validations__/0
    (vex) /tmp/vex/test/test_helper.exs:8: Vex.Struct.__before_compile__/1
```

So, as a test,  I added the following line to Vex's `UserTest` module in `test_helper.exs` and got the same error.
```elixir
validates :age, by: &(&1 > 18)
```
Any thoughts?
Cheers!

**Comment by benwilson512** [comment]
Hey, glad you're finding it helpful. I think in this case you're actually running up against VM limitations. See the discussion here for details: https://groups.google.com/forum/#!topic/elixir-lang-talk/7FXDKJ_Lpzg

Basically, you can't do:
```elixir
defmodule AnonymousFunctionTest do
  @anon fn() -> true end

  def anonymous_function,
    do: @anon
end
```

This is because anonymous functions are closures, which is to say that they capture the environment in which they're created. In the case of module attributes, this is the compile time environment that is not available at runtime.

Long story short, when validating structs you have to use named remote functions IE ```&OtherModule.function/1```

Make sense?

**Comment by benwilson512** [comment]
I should add that probably the easiest way to do that in your case is:
```elixir
defmodule User do
  #stuff
  defmodule Validations do
    def over_18(age) do: age > 18
  end
  validates :age, by: &Validations.over_18/1
end
```

**Comment by bitchef** [comment]
Hi Ben,
Problem solved! Thank you for the detailed reply. I am not very familiar with the intricacies of macros yet, but I will definitely take a look at the link.
Thanks again!

**Comment by mtwilliams** [comment]
Alternatively you can do:

```elixir
defmodule User
  validates :age, by: fn (age) -> age > 18 end
end
```

If you're really enterprising, you can even rewrite in `Vex.Structs.validates`...

```elixir
# Pseudo-code

rewrite_without_captures = fn (fun) ->
  {:&, _, body} = fun

  body_without_captures =
    # Walk branches in |body| recursively, replacing:
    # {:&, _, [1]} -> {name, [], Elixir}

  {:fn, [], [{:->, [], [[{name, [], Elixir}], body_without_captures]}]}
end

Enum.map(validations, fn (validation) ->
  case validation do
    {:by, [{:function, fun} | _] = opts} ->
      {:by, Keyword.merge(opts, [function: rewrite_without_captures(fun)])}
    {:by, fun} ->
      {:by, rewrite_without_captures(fun)}
    _ ->
      validation
  end
end
```

**Comment by benwilson512** [comment]
Locked. Please don't delete replies. Even when they're erroneous they provide opportunities to learn, and deleting them just confuses everyone. This isn't directed at you bitchef.
